### ISSUE TITLE
Feature request quick jump by file grouping

### ISSUE THREAD

**Issue by fourjay** [romainl/vim-qf#1]
I've added the following snippet to me (copied from Ag/Ack QF window list addition)
as a nicety when:

Several files matched and multiple lines matched.

The mapping allows me to jump around between file groupings.

} and { seem logical mappings to me, as file group seems similar to
"block"

function s:first_char_of_line()
    return getline('.')[0]
endfunction
function s:next_not_first_match()
    execute "normal /^[^" . <SID>first_char_of_line() . "]\<cr>"
endfunction
function s:prev_not_first_match()
    execute "normal ?^[^" . <SID>first_char_of_line() . "]\<cr>"
endfunction

" and in the QF window mapping function I have set up:

nnoremap <silent> <buffer> {  :call <SID>prev_not_first_match()<cr>
nnoremap <silent> <buffer> }  :call <SID>next_not_first_match()<cr>

**Comment by romainl** [comment]
That's indeed an interesting feature. I'll work on it this week-end (as well as a couple of bugfixes). Thank you.

**Comment by ntnn** [comment]
I did that a while ago:

    fu! QfGetFilepath(line) abort
        return substitute(a:line, "|.*$", "", "")
        "                          |      |   +- no flags
        "                          |      +- replace match with nothing
        "                          +- match from the first pipe to the end of line
        "                             declaring EOL explicitly is faster than implicitly
    endfu

    fu! QfJump(forward) abort
        let l:curpos = line('.')

        let l:curline = QfGetFilepath(getline('.'))
        let l:lastline = l:curline

        while l:curline == l:lastline
            if a:forward
                let l:curpos += 1
            else
                let l:curpos -= 1
            endif

            let l:lastline = l:curline
            let l:curline = QfGetFilepath(getline(l:curpos))
        endwhile

        call setpos('.', [bufnr('%'), l:curpos, 1, 0])
    endfu
    nnoremap <buffer> { :call QfJump(0)<cr>
    nnoremap <buffer> } :call QfJump(1)<cr>

**Comment by romainl** [comment]
@ntnn that's funny, I pushed the following last night:

```
" experimental feature
" jump to previous/next file grouping
function qf#PreviousFile()
    normal! 0
    let cfile = expand("<cfile>")
    while getline(".") =~ cfile && line(".") != 1
        normal! k
    endwhile
endfunction

function qf#NextFile()
    normal! 0
    let cfile = expand("<cfile>")
    while getline(".") =~ cfile && line(".") != line("$")
        normal! j
    endwhile
endfunction
```

which could certainly benefit from a bit of refactoring (but it was late).

Also, I think it would probably be best if those motions jumped to the top of the file grouping.

**Comment by ntnn** [comment]
That's a lot neater - you could use the `getFilePath` from my solution
to eliminate `normal! 0 | let cf...`.
For jumping to the top of a file group it'd be sufficient to go up two
times then down one:

    fu! qf#GetFilePath(line) abort
        return substitute(a:line, "|.*$", "", "")
        "                          |      |   +- no flags
        "                          |      +- replace match with nothing
        "                          +- match from the first pipe to the end of line
        "                             declaring EOL explicitly is faster than implicitly
    endfu

    fu! qf#SearchNextFileChunk(down) abort
        let cfilepath = qf#GetFilePath(getline("."))
        if a:down
            let direction = j
            let end = "$"
        else
            let direction = k
            let end = "1"
        endif

        while qf#GetFilePath(getline(".")) == cfilepath && line(".") != end
            execute 'normal! ' . direction
        endwhile
    endfu

    function qf#PreviousFile() abort
        call qf#SearchNextFileChunk(0) | qf#SearchNextFileChunk(0)
        normal! j
    endfunction

    function qf#NextFile() abort
        call qf#SearchNextFileChunk(1)
    endfunction

Also, the `&& line(".") != end` part isn't actually needed,
if the abort flag for the function is set - depends on how you want it.

I also think the `qf#GetFilePath` should still be used to get the
correct path to avoid problems where a match was found in e.g.
a makefile, resulting in a qf item starting with `makefile|` and
e.g. a match in another makefile further down, resulting in the next
file chunk starting with `third_party/makefile` etc. In that case your
solution would jump from `makefile|` over `third_party/makefile` all the
way down until something else comes.

**Comment by ntnn** [comment]
I fiddled a bit around with this and ended up with this:

    fu! qf#GetFilePath(line) abort
        return substitute(a:line, "|.*$", "", "")
        "                          |      |   +- no flags
        "                          |      +- replace match with nothing
        "                          +- match from the first pipe to the end of line
        "                             declaring EOL explicitly is faster than implicitly
    endfu

    fu! qf#JumpToFirstItemOfFileChunk() abort
        let l:chunk_file_path = qf#GetFilePath(getline('.'))

        while line('.') - 1 != 0 && l:chunk_file_path == qf#GetFilePath(getline(line('.') - 1))
            normal! k
        endwhile
    endfu

    fu! qf#JumpFileChunk(down) abort
        let l:start_file_path = qf#GetFilePath(getline('.'))
        let l:direction = a:down ? "j" : "k"
        let l:end       = a:down ? "$" : 1

        while l:start_file_path == qf#GetFilePath(getline('.')) && line(".") != line(l:end)
            execute 'normal! ' . l:direction
        endwhile

        call qf#JumpToFirstItemOfFileChunk()
    endfu

Turns out, when `normal! j/k` is used the first/last line need to be
checked, my previous solution probably silently failed because `setpos(`
threw an error, which let the function itself abort.

**Comment by romainl** [comment]
We are almost there.

I fixed a few small mistakes and the whole thing seems to work relatively well.

    function qf#GetFilePath(line) abort
        return substitute(a:line, "|.*$", "", "")
        "                          |      |   +- no flags
        "                          |      +- replace match with nothing
        "                          +- match from the first pipe to the end of line
        "                             declaring EOL explicitly is faster than implicitly
    endfunction

    function qf#SearchNextFileChunk(down) abort
        let cfilepath = qf#GetFilePath(getline("."))
        if a:down
            let direction = "j"
            let end = line("$")
        else
            let direction = "k"
            let end = 1
        endif

        while qf#GetFilePath(getline(".")) == cfilepath && line(".") != end
            execute 'normal! ' . direction
        endwhile
    endfunction

    function qf#PreviousFile() abort
        call qf#SearchNextFileChunk(0) | call qf#SearchNextFileChunk(0)
        normal! j
    endfunction

    function qf#NextFile() abort
        call qf#SearchNextFileChunk(1)
    endfunction

The only annoyance left is that the last `{` ends up on the second line.

**Comment by romainl** [comment]
@ntnn, I pulled in your last snippet and it seems to work reliably. I'll wait a bit before tagging a new release, though. Just in case.

**Comment by romainl** [comment]
Fixed as of commit 9df30b26d349dff80136ed401e164fe5ff92b6cb.
